lib/active_model/serializer/collection_serializer.rb

```diff
@@ -45,7 +45,14 @@ def json_key
         # 2. get from first serializer instance in collection
         key ||= (serializer = serializers.first) && serializer.json_key
         # 3. get from collection name, if a named collection
-        key ||= object.respond_to?(:name) ? object.name && object.name.underscore : nil
+        key ||= begin
+                  object_name = object.respond_to?(:name) ? object.name : nil
+                  if (object_name && object_class = object_name.safe_constantize)
+                    serializer_class = serializer_from_resource(object_class.new, ActiveModel::Serializer, options) { nil }
+                  end
+                  serializer_class.nil? ? (object_name && object_name.underscore) : serializer_class.json_key
+                end
+
```

Comment:
Sorry -- was fighting an uphill battle with rubocop's `Metrics/PerceivedComplexity`. Based on your comment line 38 I figured it might be okay to disable that for now? Let me know if that's misguided and you'd prefer me to break out into other methods.